### PR TITLE
.NET rewrite

### PR DESCRIPTION
**Pull request by dealloc**
as discussed in #28 the API will be rewritten from Elixir to C#
The main goal is to provide a language/ecosystem that is more accessible to contribution, part of this effort is also the [helldivers-2/json](https://github.com/helldivers-2/json) repository.

This PR switches over the Elixir implementation to an [AOT](https://learn.microsoft.com/en-us/dotnet/core/deploying/native-aot/?tabs=net7%2Cwindows) ASP.NET Core application written in C#

### Goals
- lower memory footprint by keeping as much shared memory
- fast startup (see below)
- faster (AOT, no reflection and compiler optimization)
- higher API limits
- extensible
- backwards compatible with ArrowHead's official API
- approachable for new contributors

### Faster, better, stronger
So why is faster startup and lower memory usage through AOT such a big deal?
It basically comes down to the fact that the API is currently hosted on my own costs, and to keep those down as much as possible I try to minimize the KPI's most provides bill for: CPU time and memory usage.

By using less RAM and less CPU I can host the API on a lower tier, saving me costs and allowing me to put in higher API rate limits

### Architecture
The new API is divided up in 4 components to give each part (domain) of the application it's own 'space':
- Helldivers-2-API: contains the actual HTTP API interface, OpenAPI documents and also serves as the runtime host (aka the application that starts)
- Helldivers-2-Models: contains POCO (_P_ lain _O_ ld _C_ # _O_ bjects) of both the ArrowHead API and the 'new' API surface
- Helldivers-2-Core: contains the 'snapshot' of information currently being served and mapping from ArrowHead models to domain models
- Helldivers-2-SourceGen: I admit I may have gone a little overboard, see below)

### Source generators
As mentioned, the 4th component is `Helldivers-2-SourceGen`, which basically does 1 thing:
it generates a class `Helldivers.Models.Static` with properties for every JSON file pulled in from [helldivers-2/json](https://github.com/helldivers-2/json).

Generate source files look a lot like this:
```csharp
// <auto-generated />
#pragma warning disable CS1591 // Missing XML comment for publicly visible type or member
using global::System.Collections.Generic;

namespace Helldivers.Models;

public static partial class Static
{
    /// <summary>Public list of Factions entries from factions.json</summary>
    public static IReadOnlyDictionary<int, string> Factions = new Dictionary<int, string>()
	{
		{ 1, "Humans" },
		{ 2, "Terminids" },
		{ 3, "Automaton" },
	};
}
```

This allows us to [intern](https://learn.microsoft.com/en-us/dotnet/api/system.string.intern?view=net-8.0) names of sectors, planets etc for even more efficient memory usage.
It also makes sure I don't have to manually write out the names of planets and sectors in the C# code, or load them in at runtime (see 'Faster, better, stronger')

### Migration path
if you are using the current API endpoints from the Elixir version, there will be *breaking changes*.
Please check the OpenAPI documents under `Helldivers-2-API/wwwroot/` for details on the new endpoints.
Also feel free to add comments below here for endpoints you are using, and to see how those can be migrated in the new model.

### Roadmap
- [X] map ArrowHead API endpoints
- [X] implement rate limits
- [X] add API endpoints for fetching specific domain data
- [X] map attacks, events and campaigns
- [X] add documentation
- [X] ~~document language selection~~ scrapped for later date
- [X] add contribution guide
- [X] update CI for .NET
- [X] update CD for .NET
- [X] ~~generate documentation for .NET rather than Elixir~~ scrapped for later date
- [x] publish generated documentation as Github pages site

### CODE REVIEW

**Comment by lambstream** [comment]
🎊

**Review by lambstream — DISMISSED** [review]
Looks good to me. 

We should add a task or an issue for the handling of the language detection. (As mentioned in the comments of the ArrowHeadController.

**Comment by dealloc** [comment]
@chatterchats can I ask for your eyes on this PR as well?
Specifically the documentation part around contribution etc, though you're free of course to review other things you'd like as well

**Comment by chatterchats** [comment]
Contribution guide looks good, will be a lot better with the todos are finished there but it's great to get started!

File: src/Helldivers-2-Core/Facades/V1Facade.cs [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+    IStore<Campaign, int> campaignStore,
+    CampaignMapper campaignMapper,
+    IStore<Assignment, long> assignmentStore,
+    AssignmentMapper assignmentMapper,
+    IStore<Dispatch, int> dispatchStore,
+    DispatchMapper dispatchMapper
+)
+{
+    /// <see cref="IStore{T,TKey}.SetStore" />
+    public async ValueTask UpdateStores(Models.ArrowHead.WarId warId, Models.ArrowHead.WarInfo warInfo, Dictionary<string, Models.ArrowHead.WarStatus> warStatuses, Models.ArrowHead.WarSummary warSummary, Dictionary<string, List<Models.ArrowHead.NewsFeedItem>> newsFeeds, Dictionary<string, List<Models.ArrowHead.Assignment>> assignments)
```
Comment (lambstream) [inline src/Helldivers-2-Core/Facades/V1Facade.cs:25]: We should probably map the warId here at some point.

File: src/Helldivers-2-API/Middlewares/RateLimitMiddleware.cs [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+using System.Threading.RateLimiting;
+
+namespace Helldivers.API.Middlewares;
+
+/// <summary>
+/// Handles applying rate limit logic to the API's requests.
+/// </summary>
+public sealed partial class RateLimitMiddleware(ILogger<RateLimitMiddleware> logger, IMemoryCache cache) : IMiddleware
+{
+    private const int DefaultRequestLimit = 5;
```
Comment (lambstream) [inline src/Helldivers-2-API/Middlewares/RateLimitMiddleware.cs:12]: These should be changed to .env variables.

We would also like to have separate rate limit policies.

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+dotnet restore
+dotnet build --no-restore --no-incremental
+```
+If you don't get any errors, that means your development environment is all set to get developing!
+Open the `Helldivers-2.sln` file in the root of the repository and you're good to go!
+
+### Formatting, line endings and syntax
+There's three pilars when it comes to coding conventions, formatting etc:
+- [.editorconfig](https://editorconfig.org/) is set up in the repository, if your IDE supports it you should install
+    the required plugin to ensure your code is correctly formatted according to it.
```
Comment (lambstream) [inline CONTRIBUTING.md:36]:
```suggestion
    the required plugin to ensure your code is correctly formatted according to the configuration.
```

File: CONTRIBUTING.md [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+- [Microsoft C# coding conventions](https://learn.microsoft.com/en-us/dotnet/csharp/fundamentals/coding-style/coding-conventions)
+    write idiomatic C# and try to follow conventions from either Microsoft or the project if it deviates from a specific
+    rule.
+- [dotnet-format](https://github.com/dotnet/format) we use `dotnet format` in our CI pipeline to ensure consistent
+    formatting, line endings and syntax. Before submitting a pull request, make sure you run `dotnet format` on your code.
+
+### Be respectful
+Disagreements happen, discussions are perfectly fine (we encourage a healthy discussion even!), however you (and we) are
+to remain respectful at **all times**. While we try to listen to the community whenever possible, please keep in mind
+that 1. we do this in our free time (and for free nonetheless), and 2. we have the *final* say when it comes to decisions
```
Comment (lambstream) [inline CONTRIBUTING.md:46]:
```suggestion
that 1. we do this in our spare time (for free, to boot!), and 2. we have the *final* say when it comes to decisions
```

File: CONTRIBUTING.md [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+## Why we might not accept all Pull Requests
+There are three primary reasons we might choose not to accept your pull request (however it's not limited to these two);
+1. **We don't accept pull requests that are cosmetic in nature** (think formatting, whitespace, ...)
+    we don't do this for the same reason the [Ruby Rails](https://github.com/rails/rails) project does not allow such
+    changes, [TLDR](https://github.com/rails/rails/pull/13771#issuecomment-32746700) it adds noise and makes maintenance
+    even more difficult than it already is.
+2. **We don't accept pull requests with features we don't want/feel comfortable maintaining**
+    If you create an awesome feature that is rather complex, we might choose not to accept the feature.
+    This is probably not related to the quality of your code (that's what PR's are for in the first place), but rather
+    with the fact that after you've left the maintainers still need to make sure it keeps working for as long as we
```
Comment (lambstream) [inline CONTRIBUTING.md:67]:
```suggestion
    with the fact that after you've left, the maintainers still need to make sure it keeps working for as long as we
```

**Review by lambstream — APPROVED** [review]
Looks good to me.

**Pull request merged**